### PR TITLE
Feature Detection and Descriptor Classes

### PR DESCRIPTION
**Pull request by Cryptum169**
## Additions
- Feature Extractor Abstract Class
- SIFT, ORB and Superpoint Feature detector and descriptor
- Corresponding pyTests

## Example Output
**ORB**
Screenshot from 2021-01-23 19-40-07

**SIFT**
Screenshot from 2021-01-23 19-41-08

**Superpoint**
Screenshot from 2021-01-23 19-42-16

## Possibly Breaking Changes
- Split current configuration files into Dataloader config files and feature extractor config files.
- Tracker class takes in additional optional `FeatureExtractor` argument for detection

### CODE REVIEW

**Comment by Cryptum169** [comment]
A couple of issues might need more discussion:

1. `FeatureExtraction` class works by being inside a `tracker` class, which could be problematic that it'd require an independent `Tracker` object for every different feature. But the downside being that Feature detection and tracking is kind of incorporated within `Tracker`, maybe not a wise idea to touch it in this PR.
2. ORB's implementation of max number of keypoints used a OctTree for picking the best one, while openCV straight up sorted and picked the first `n` keypoints with largest responses in every pyramid level. Doesn't seem to affect the result much.
3. Python for loops are kind of ... slow? Runs at about 200ms. Roughly evenly split between image pyramid computation and keypoint computation.

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "vslam/feature/SuperPointPretrainedNetwork"]
+	path = vslam/feature/SuperPointPretrainedNetwork
```
Comment (akashsharma02) [inline .gitmodules]: Maybe create a thirdparty folder and add this there?

File: test/test_feature.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        
+        for i in tqdm.tqdm(range(1, len(tum)), desc="TumDataloader"):
+            prev_rgb = tum[i - 1]['rgb']
+            curr_rgb = tum[i]['rgb']
+            
+            T_cw_orig, _, _, _, _ = native.bootstrap(prev_rgb, curr_rgb)
+            T_cw_sift, _, _, _, _ = sift_tracker.bootstrap(prev_rgb, curr_rgb)
+            T_cw_orb, _, _, _, _ = orb_tracker.bootstrap(prev_rgb, curr_rgb)
+
+            err_sift = np.linalg.norm(T_cw_orig.between(T_cw_sift).translation())
```
Comment (akashsharma02) [inline test/test_feature.py:57]: I wonder if we should change the return type of bootstrap to SE3Pose wrapper or just use gtsam.Pose3 and delete the se3pose.py file in vslam/types. What do you think?

File: vslam.py [hunk 3]
```diff
@@ -50,6 +45,10 @@ def main(config):
                         default=None,
                         help="Path to the configuration yaml file",
                         type=str)
-
+    parser.add_argument("-f",
```
Comment (akashsharma02) [inline vslam.py]: I think it's better to have a single configuration file. Please add feature as a component in the same config file

File: vslam/feature/feature.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        pass
+
+    @abstractmethod
+    def detect(self, img: np.ndarray) -> list:
+        """
+
+        Args:
+            Image in the form of ndarray
+        Returns:
+            Interest points
```
Comment (akashsharma02) [inline vslam/feature/feature.py]: can return np array of points Nx2 ndarray?

File: vslam/feature/feature.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    @abstractmethod
+    def compute(self, img: np.ndarray, feat: np.ndarray) -> Tuple[list, np.ndarray]:
+        """Compute Descriptors with given interest points
+
+        Args:
+            Image
+            Interest Points
+
+        Returns:
+            Tuple of interest points and descriptors
```
Comment (akashsharma02) [inline vslam/feature/feature.py]: same return Nx2 ndarray of interest points, and NxD ndarray of corresponding descriptors

File: vslam/feature/feature.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    @abstractmethod
+    def detectAndCompute(self, img: np.ndarray) -> Tuple[list, np.ndarray]:
+        """Detect Interest Points and compute corresponding feature descriptors
+
+        Args:
+            Image
+
+        Returns:
+            Tuple of Keypoint and Corresponing Descriptors
```
Comment (akashsharma02) [inline vslam/feature/feature.py]: same as above.

File: vslam/feature/orb_feature.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    """
+    def __init__(self, **kwargs) -> None:
+        """Contructor
+            Args: 
+        """
+        super().__init__()
+        self.orb = cv2.ORB_create(**kwargs)
+
+        self.nlevels = self.orb.getNLevels()
+        self.scaleFactor = self.orb.getScaleFactor()
```
Comment (akashsharma02) [inline vslam/feature/orb_feature.py]: all variables are snake_case and all functions are camelCase, and all classes are PascalCase

File: vslam/feature/orb_feature.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        """Wrapper class for ORB detectAndCompute
+            Args:
+                Image
+                Mask
+
+            Returns:
+                Tuple of Keypoint and descriptors
+        """
+        img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+        imgPyrd = self._computePyramid(img)
```
Comment (akashsharma02) [inline vslam/feature/orb_feature.py]: I wonder why this necessary? Doesn't orb detectAndCompute already detect features at different scales?

I think this code is still useful for when we write ORB2FeatureExtractor. What do you think?

File: vslam/feature/superpoint_feature.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        """Wrapper class for SIFT compute
+            Args:
+                Image
+
+            Returns:
+                Tuple of keypoint and corresponding feature descriptor
+        """
+        pass
+
+    def _pts_to_keypoints(self, pts : np.ndarray, orig_shape : Tuple, resized_shape : Tuple) -> List:
```
Comment (akashsharma02) [inline vslam/feature/superpoint_feature.py:53]: maybe makes sense to add these types of functions into a separate module as vslam utils? this function is also written in Tracker as far as I remember.

File: vslam/parse_config.py [hunk 10]
```diff
@@ -35,10 +35,20 @@ def from_args(cls: ConfigParser, args: argparse.ArgumentParser) -> ConfigParser:
...
+
+        msg_no_feat = "Configuration file needs to be specified. Add '-f feat.yaml' argument when running"
+        assert args.feat is not None, msg_no_feat
+
         cfg_filename = Path(args.config)
 
         filestream = open(cfg_filename, 'r')
         config = yaml.load(filestream, Loader=yaml.FullLoader)
+
+        feat_filename = Path(args.feat)
```
Comment (akashsharma02) [inline vslam/parse_config.py]: same comment as earlier. Better to have a single file configuration

File: vslam/tracker.py [hunk 11]
```diff
@@ -193,39 +195,43 @@ def matchImagePair(
...
-        prev_keypoints, prev_desc = orb.compute(prev_image_gray,
-                                                prev_keypoints)
-        curr_keypoints, curr_desc = orb.compute(curr_image_gray,
-                                                curr_keypoints)
+        # If feature provided
+        if self.feature is not None:
+            prev_keypoints, prev_desc = self.feature.detectAndCompute(prev_image)
+            curr_keypoints, curr_desc = self.feature.detectAndCompute(curr_image)
+        else:
+            # 1. Detect keypoints in image 1 (prev) and 2 (curr)
```
Comment (akashsharma02) [inline vslam/tracker.py]: else use a default feature extractor say KLTFeatureExtractor? and move this code there?

**Review by akashsharma02 — CHANGES_REQUESTED** [review]
Thanks for your PR. This is great, I've suggested some minor changes. Mostly style issues:

please use snake_case for variables, camelCase for function names, PascalCase for classes

File: vslam.py [hunk 12]
```diff
@@ -8,6 +8,7 @@
 from vslam.types import PinholeCamera, Frame, Map
 from vslam.tracker import Tracker
 from vslam import feature
+from vslam import feature
```
Comment (akashsharma02) [inline vslam.py]: duplicate line?

File: vslam/feature/feature.py [hunk 13]
```diff
@@ -11,37 +11,38 @@ def __init__(self, **kwargs) -> None:
...
-    def detectAndCompute(self, img: np.ndarray) -> Tuple[list, np.ndarray]:
+    def detectAndCompute(self, img: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
         """Detect Interest Points and compute corresponding feature descriptors
 
         Args:
             Image
 
         Returns:
-            Tuple of Keypoint and Corresponing Descriptors
+            Tuple of interest points and corresponing descriptors
```
Comment (akashsharma02) [inline vslam/feature/feature.py]: typo at corresponding

File: vslam/feature/sift_feature.py [hunk 14]
```diff
@@ -14,17 +14,17 @@ def __init__(self, **kwargs) -> None:
         super().__init__()
         self.sift = cv2.SIFT_create(**kwargs)
 
-    def detect(self, img: np.ndarray, mask: np.ndarray = None) -> list:
+    def detect(self, img: np.ndarray, mask: np.ndarray = None) -> np.ndarray:
```
Comment (akashsharma02) [inline vslam/feature/sift_feature.py:18]: if it's a list, type annotation must be List with (include from typing import List)

File: vslam/feature/orb_feature.py (hunk 7 above)
Comment (akashsharma02) [reply vslam/feature/orb_feature.py]: okay I'll change this after merge

File: vslam/feature/orb_feature.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+    def compute(self, img: np.ndarray, kp: list) -> Tuple[list, np.ndarray]:
+        """Wrapper class for ORB compute
+            Args:
+                Image
+
+            Returns:
+                Tuple of keypoint and corresponding feature descriptor
+        """
+        img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+        return self.orb.compute(img, kp)
```
Comment (akashsharma02) [inline vslam/feature/orb_feature.py]: shouldn't self._computeKeypoints be called here, if you are implementing it differently?

File: vslam/feature/orb_feature.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+
+    def _runByImageBorder(self, kps: list, imgSize : Tuple) -> List:
+        """Remove keypoints too close to the image border
+            Args:
+                Keypoints
+                Image Size
+                EdgeThreshold
+
+            Return:
+                Keypoints not immediatley close to image border
```
Comment (akashsharma02) [inline vslam/feature/orb_feature.py]: typo

**Review by akashsharma02 — APPROVED** [review]
I've approved the changes for merge, but minor typos and style issues still remain.
Not major but would appreciate changes.

File: vslam/feature/orb_feature.py (hunk 15 above)
Comment (Cryptum169) [reply vslam/feature/orb_feature.py]: Wait sorry I'm not done changing this yet 😬 I had to change to a different machine. I'm changing this back to simply grid-ing the image and do detect from there.

**Comment by Cryptum169** [comment]
@akashsharma02 Frame object made a call to `detectAndCompute(img, num_interest_points)` [Link](https://github.com/akashsharma02/visual-slam/blob/0ed707890f7b6098b2c42674be68837b84cab36f/vslam/types/frame.py#L58)
Can we remove the argument of `num_interest_points` since this is typically specified in the constructor of feature extractor configs?

**Comment by Cryptum169** [comment]
I have fixed the remaining typos and style inconsistencies, sorry it took longer than expected.

**Review by akashsharma02 — APPROVED** [review]

**Pull request merged**